jsx-factory.js

```diff
@@ -0,0 +1,30 @@
+/**
+ * @typedef {import('./lib/core.js').Element} Element
+ * @typedef {import('./lib/core.js').Root} Root
+ * @typedef {import('./lib/core.js').HResult} HResult
+ * @typedef {import('./lib/core.js').HArrayChild} HArrayChild
+ * @typedef {import('./lib/core.js').HProperties} HProperties
+ * @typedef {import('./lib/core.js').HPropertyValue} HPropertyValue
+ * @typedef {import('./lib/core.js').HStyle} HStyle
+ * @typedef {import('./lib/core.js').core} Core
+ *
+ * @typedef {Record<string, unknown>} JSXProps
+ * @typedef {(props: JSXProps) => HResult } HFn
+ */
+
+import {h as hast} from './index.js'
+
+/**
+ * @param {string | null | HFn} typeOrFn
+ * @param { JSXProps | undefined } props
+ * @param { HArrayChild } children
+ * @returns HResult
+ */
+export function h(typeOrFn, props, ...children) {
+  if (typeof typeOrFn === 'function') {
+    return typeOrFn({...props, children})
```

Comment:
I think it would be nice to infinitely flatten children, so we can pass this in JSX:

```jsx
function Foo({ children }) {
  // children is ['bar', 'baz', 'fooz']
}

<Foo>
  {'bar'}
  {['baz']}
  {[[[['fooz']]]]}
<Foo>
```

This is just a silly hardcoded example, but in practice I think it’s common to use a lot of `.map()` and `.filter()` with JSX.

Reply:
@remcohaszing Do you think this should be a blocker?

Reply:
Yes, because changing it later would be a breaking change.

Reply:
* There are cases where `children` is currently not an array, right? (i saw a test for a function). Do we even want to support that?
* if anything can be passed (such as functions), that also means a user can pass an array by reference, which we shouldn’t touch, how to handle that?

Reply:
I think children is just an iterable of values that have meaning to the function. If that function creates an intrinsic element like `A`, then it makes no sense to have children that are anything other than child nodes.  However, if it is a function component, then it's up to the function how it wants to interpret its arguments.

I would say you definitely do _not_ want to flatten children in the case of a component, but it would be an option for creating low level html nodes. I don't have enough experience with what flattening would feel like to have an opinion.

Reply:
The thing is, in this example, `Component` gets a `children` prop as a string.

```jsx
<Component>
  {'foo'}
</Component>
```

In this example, it’s an array of two strings.

```jsx
<Component>
  {'foo'}
  {'bar'}
</Component>
```

React normalizes this. I think it’s a good idea to do this here as well. It makes sense to me to always pass a flat array, but I’m open to other suggestions as well. Perhaps strings should even be transformed to text nodes?

Reply:
> React normalizes this.

That’s not what React does, it’s how Babel and others do their JSX compilation. And there’s as way around it: by passing `children`. `children` is just a prop. `<X children={SomeFunctionOrClassOrWhatever} />` or so is fine.

Some examples: https://babeljs.io/repl#?browsers=defaults%2C%20not%20dead&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=DwQQBAxgFglgNgEwE4FMB2BeA3gbQEQCGeANGAIwC6AvgHy54BGJYATNfRMwMykQD2aAM584KAHRw-Ac2rAA9CBpA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.22.5&externalPlugins=&assumptions=%7B%7D

---

I don’t think we should touch `children`. a) React and similar don’t either, b) it would be slow to iterate / call `.flat()` on every children array?